core/interop/src/lib.rs

```diff
@@ -301,6 +304,50 @@ impl<T: TryFromJs> TryFromJsArgument for JsThis<T> {
     }
 }
 
+/// Captures a [`ContextData`] data from the [`Context`] as a JS function argument,
+/// based on its type.
+///
+/// The host defined type must implement [`Clone`], otherwise the borrow
+/// checker would not be able to ensure the safety of the context while
+/// making the function call. Because of this, it is recommended to use
+/// types that are cheap to clone.
+///
+/// For example,
+/// ```
+/// # use boa_engine::{Context, Finalize, JsData, JsValue, Trace};
+/// use boa_interop::{IntoJsFunctionCopied, ContextData};
+///
+/// #[derive(Clone, Debug, Finalize, JsData, Trace)]
+/// struct CustomHostDefinedStruct {
+///    #[unsafe_ignore_trace]
+///    pub counter: usize,
+/// }
+/// let mut context = Context::default();
+/// context.insert_data(CustomHostDefinedStruct { counter: 123 });
+/// let f = (|ContextData(host): ContextData<CustomHostDefinedStruct>| {
+///   host.counter + 1
+/// }).into_js_function_copied(&mut context);
+///
+/// assert_eq!(f.call(&JsValue::undefined(), &[], &mut context), Ok(JsValue::new(124)));
+/// ```
+#[derive(Debug, Clone)]
+pub struct ContextData<T: Clone>(pub T);
+
+impl<'a, T: NativeObject + Clone> TryFromJsArgument<'a> for ContextData<T> {
+    fn try_from_js_argument(
+        _this: &'a JsValue,
+        rest: &'a [JsValue],
+        context: &mut Context,
+    ) -> JsResult<(Self, &'a [JsValue])> {
+        match context.get_data::<T>() {
+            Some(value) => Ok((ContextData(value.clone()), rest)),
+            None => Err(JsError::from_opaque(
+                js_string!("Host defined value not found").into(),
+            )),
```

Comment:
Would probably be better to create a `TypeError` instead.

Reply:
Done.